pkg/policymutation/policymutation.go

```diff
@@ -390,68 +394,80 @@ func GeneratePodControllerRule(policy kyverno.ClusterPolicy, log logr.Logger) (p
 // - "none" if:
 //          - name or selector is defined
 //          - mixed kinds (Pod + pod controller) is defined
+//          - Pod and PodControllers are not defined
 //          - mutate.Patches/mutate.PatchesJSON6902/validate.deny/generate rule is defined
 // - otherwise it returns all pod controllers
 func CanAutoGen(policy *kyverno.ClusterPolicy, log logr.Logger) (applyAutoGen bool, controllers string) {
+	var needAutogen bool
 	for _, rule := range policy.Spec.Rules {
 		match := rule.MatchResources
 		exclude := rule.ExcludeResources
 
 		if match.ResourceDescription.Name != "" || match.ResourceDescription.Selector != nil || match.ResourceDescription.Annotations != nil ||
 			exclude.ResourceDescription.Name != "" || exclude.ResourceDescription.Selector != nil || exclude.ResourceDescription.Annotations != nil {
 			log.V(3).Info("skip generating rule on pod controllers: Name / Selector in resource description may not be applicable.", "rule", rule.Name)
-			return false, "none"
+			return false, ""
 		}
 
 		if isKindOtherthanPod(match.Kinds) || isKindOtherthanPod(exclude.Kinds) {
-			return false, "none"
+			return false, ""
 		}
 
+		needAutogen = hasAutogenKinds(match.Kinds)
+
 		for _, value := range match.Any {
 			if isKindOtherthanPod(value.Kinds) {
```

Comment:
Looks like this check is missing from `any/all.exclude.kinds` as you pointed out here https://github.com/kyverno/kyverno/blob/9e10eef42203b92d5d3c94a30a378d89c46ab29b/pkg/policymutation/policymutation.go#L406-L408, do we need to add it?

Reply:
See https://github.com/kyverno/kyverno/pull/2737#discussion_r752318568

Reply:
Quoting seh,
> This is odd. If I write a policy that does not apply to Pods or, say, Deployments by way of exclusion, why does that induce automatic generation of rules?

Reply:
When excluding `Pod`, by default additional auto-gen rules should be added to exclude pod controllers, right?

Reply:
I think, not sure though

Reply:
Do you have any exceptions in mind? If users do not want this behavior, they can tune/turn it off by configuring the annotation.

Reply:
I don't but makes sense. I'll make the changes.

Reply:
There are two cases for matching:
1. We match a mixed kind, say, `{"Pods", "Secrets", "Deployments"}`: `isKindOtherthanPod()` returns true and autogen is disabled
2. We match only one of the Pod Controllers

In Case 2, any exclude rule would make sense only if it applies to a subset of matched resources. Hence only one of the Pod Controllers. So even if we enable autogen, it would only create extra rules which are redundant since, match kinds and exclude kinds would be disjoint. Here's an example:

Rule
```yaml
  - name: validate-registries
    match:
      all:
      - resources:
          kinds:
          - Deployment
    exclude:
      any:
        - subjects:
           - kind: Pod
             name: test
```

Rules after autogen is applied:
```yaml
  rules:
  - exclude:
      any:
      - resources: {}
        subjects:
        - kind: Pod
          name: test
      resources: {}
    generate:
      clone: {}
    match:
      all:
      - resources:
          kinds:
          - Pod
      resources: {}
    mutate: {}
    name: validate-registries
    validate:
      message: Images may only come from our internal enterprise registry.
      pattern:
        metadata:
          name: test-ns
  - exclude:
      any:
      - resources: {}
        subjects:
        - kind: Pod
          name: test
      resources: {}
    generate:
      clone: {}
    match:
      all:
      - resources:
          kinds:
          - DaemonSet
          - Deployment
          - Job
          - StatefulSet
      resources: {}
    mutate: {}
    name: autogen-validate-registries
    validate:
      message: Images may only come from our internal enterprise registry.
      pattern:
        spec:
          template:
            metadata:
              name: test-ns
  - exclude:
      any:
      - resources: {}
        subjects:
        - kind: Pod
          name: test
      resources: {}
    generate:
      clone: {}
    match:
      all:
      - resources:
          kinds:
          - CronJob
      resources: {}
    mutate: {}
    name: autogen-cronjob-validate-registries
    validate:
      message: Images may only come from our internal enterprise registry.
      pattern:
        spec:
          jobTemplate:
            spec:
              template:
                metadata:
                  name: test-ns
```

Reply:
```
  - name: validate-registries
    match:
      all:
      - resources:
          kinds:
          - Deployment
    exclude:
      any:
        - subjects:
           - kind: Pod
             name: test
```

Is this even a valid rule? `Pod` and `Deployment` do not share same spec.

Reply:
`match.all.kinds: ["Deployment"]` enables the auto-gen. `subjects` was a bad example

Reply:
>`match.all.kinds: ["Deployment"]` enables the auto-gen

What is the annotation value?

Reply:
It's the default, a list of pod controllers

Reply:
Do we generate additional rules for other pod controllers? Or the auto-gen rule also includes `Deployment`?

Reply:
We generate rules for pod controllers other than Deployment

Reply:
Is this resolved?